Fix for loading wasm files under Node.js and in browser when files we…

nazar-pc opened the pull request:
I think this is related to #4542 and, probably, fixes it.

I have compiled a library with `emcc supercop.c #files -o supercop.js -O2 --closure 1 -s WASM=1` and got 2 files in the root of the project: `supercop.js` and `supercop.wasm`, everything worked fine so far. However, when I published it to NPM effective path to WebAssembly file become `node_modules/supercop.wasm/supercop.wasm` and `supercop.js` file was unable to load it.

With this patch WebAssembly files that are located next to JavaScript files are loaded correctly both in Node.js and in browser.

I've run some tests and on the surface it looks fine, however, my 12-thread Core i7 is not enough to run all of the tests in any sane amount of time, so... Also I'm a bit terrified with number of tests being runned and do not quite understand how to add new tests. Hopefully someone can augment this PR with new tests.

kripken commented:
Some questions:

 * Does the `fetch` API exist in node.js?
 * What is a concrete use case that shows the problem? (I'm not sure what publishing to npm entails, that you mentioned.) Looking for simple steps to reproduce the failure.

nazar-pc commented:
`fetch()` is for browser only, so it doesn't matter whether Node supports it in this case.

Here is the project I've mentioned: https://github.com/nazar-pc/supercop.wasm
As you can see there is `supercop.wasm` in the root of the project, but when I install this package as dependency it is not in working directory anymore, but rather in `node_modules/supercop.wasm` subdirectory, thus `supercop.wasm` loading fails. With this patch it searches for `supercop.wasm` in the same directory as `supercop.js` rather than in current directory.

You can reproduce an issue by modifying `supercop.wasm` dependency in https://github.com/nazar-pc/bittorrent-dht-store-keypair/blob/browser-compatibility/package.json from `^3.0.1` to strictly `3.0.0` (which was compiled without fix from this PR) and run tests afterwards.

juj commented:
This code now looks like it's going down the route that we wanted to avoid in PR #4942. That is, instead of having multiple locations in code that reference to `__dirname` or do complex statements such as `document.currentScript.src.split('/').slice(0, -1).join('/');`, it would be good to have a `Module['scriptDirectory']` variable which does this only once, and all places of use would reference that variable. That would make handling this PR and #4942 look more or less identical. @nazar-pc : what do you think?

nazar-pc commented:
I think having it in one place is a good idea overall, especially for some long ones like those needed in browser environment.
Also I think this PR is basically fixing another aspect of the same root issue as #4942, namely that the files are loaded from current working directory instead of the same directory as JavaScript file.

I also think we should possibly implement common path in either PR (or even yet another one) and then rebase against that.

Should I try to implement `Module['scriptDirectory']` in this PR or in another one or will someone tackle this separately? I'd like to upstream fixes as soon as possible because relying on patched libraries and tooling is quite painful in my experience.

kripken commented:
Can implement it in this PR, I think. Perhaps ask in the other PR if someone is working on it to be sure, but I don't think anyone is (last update there was over a month ago).

RReverser commented:
Yeah, I got "distracted" from this area for the last month by other projects, so would be happy to see anyone implementing `scriptDirectory` suggestion.

nazar-pc commented:
I think this can be called **The ultimate fix**.

I've found that over time Emscripten collected a few options that were fixing this exact issue for different types of files, namely:
* memoryInitializerPrefixURL
* pthreadMainPrefixURL
* cdInitializerPrefixURL
* filePackagePrefixURL

At least some of them (I've looked through some PRs that implemented those options) originated from the fact that people were moving files around and files stopped loading.

In fact, almost always you want to load all of the files from the same directory where `.js` file is placed simply because this is how building is done by default.

Now `Module['locateFile']` will be defined in runtime if not specified before module loading. Newly introduced `Module['scriptDirectory']` is used in default built-in `Module['locateFile']` implementation.

This have one major implication: `Module['locateFile']` is now always defined, so I've decreased its priority comparing to ``Module['filePackagePrefixURL']``, ``Module['memoryInitializerPrefixURL']``, ``Module['cdInitializerPrefixURL']`` and ``Module['filePackagePrefixURL']``, so that if any of these are specified, `Module['scriptDirectory']` is not used for that particular use case.

I didn't deprecate ``Module['filePackagePrefixURL']``, ``Module['memoryInitializerPrefixURL']``, ``Module['cdInitializerPrefixURL']`` or ``Module['filePackagePrefixURL']`` yet, but chances are that even if you need to change them, they will all be the same, so effectively all you need is just new `Module['scriptDirectory']`(we can fallback to them during transition period).

Module documentation is also updated with new option (interestingly, not all of the mentioned options are mentioned there:))

This also makes #4942 unnecessary.

nazar-pc commented:
Check this carefully, it might break some test with edge-cases, but shouldn't affect any real-world application.

nazar-pc commented:
> One option might be to add a warning if they "collide".

If they collide, more specific option wins, so `Module['locateFile']` and `Module['scriptDirectory']` are always used as last resort.

> Another is not to allow setting both prefixes and locateFile.

Again, this is still supported, in this case `Module['locateFile']` is ignored completely.

The only case when it might fail is there is a test that deliberately defined both `Module['locateFile']` and, for instance, `Module['filePackagePrefixURL']`. In this case `Module['filePackagePrefixURL']` will win, while previously `Module['locateFile']` was superior. But I'd argue new behavior is more correct actually.

nazar-pc commented:
@juj, any objections here?

juj commented on ChangeLog.markdown:
```diff
@@ -14,6 +14,9 @@ Current trunk code
     - Emscripten-Clang: https://github.com/kripken/emscripten-fastcomp-clang/compare/1.37.13...incoming
+ - Added new html setting Module['scriptDirectory'] that allows customizing the URL where .wasm, .mem and some other files are located (defaults to the same location as .js file)
+ - If Module['locateFile'] is not specified explicitly it will be defined automatically using Module['scriptDirectory']
+ - Module['locateFile'] priority comparing to other options decreased since it is now always defined
```
The format that is used in the ChangeLog is to first have the changes, and then at the end of each changes to have the block

```
  - To see a list of commits in the active development branch 'incoming', which have not yet been packaged in a release, see
     - Emscripten: https://github.com/kripken/emscripten/compare/1.37.13...incoming
     - Emscripten-LLVM: https://github.com/kripken/emscripten-fastcomp/compare/1.37.13...incoming
     - Emscripten-Clang: https://github.com/kripken/emscripten-fastcomp-clang/compare/1.37.13...incoming
```

so these would go four lines up.

nazar-pc replied:
Fixed

juj commented:
Demoting `Module['locateFile']` to have different priority over the other methods would be a breaking change. Also the semantics of `Module['locateFile']` will change from being a user defineable override to having new semantics that one will need to adhere to. For example, I have instructed developers in the past to write these types of `locateFile` functions:

```
Module['locateFile'] = function(url) { return url + '.gz'; }
```

to set up gzip compression for all assets, and this new `Module['scriptDirectory']` would be ignored on all these uses. We should not change the semantics after shipping, unless there is a good reason. This does not seem like there is a need to do so?

> I didn't deprecate Module['filePackagePrefixURL'], Module['memoryInitializerPrefixURL'], Module['cdInitializerPrefixURL'] or Module['filePackagePrefixURL'] yet

I'm confused that you mention about deprecating these, but still changed to bump the priority of these to take over `Module['locateFile']`?

My understanding was that the consensus was that we don't ever want to load the `.mem`, `.wasm` etc. files from node.js's CWD, but always relative to the location of the main `.js` file. If this was solved according to https://github.com/kripken/emscripten/pull/4942/files#r113408121, then the semantics or priority of `Module['locateFile']` would not change, and relative loading would automatically get fixed for node.js developers as well who use `Module['locateFile']`. That seems superior to me, since it preserves existing functionality and fixes the relative path loading thing only that is the root bug?

nazar-pc commented:
Here is what was before:
```javascript
            var memoryInitializer = '%s';
            if (typeof Module['locateFile'] === 'function') {
              memoryInitializer = Module['locateFile'](memoryInitializer);
            } else if (Module['memoryInitializerPrefixURL']) {
              memoryInitializer = Module['memoryInitializerPrefixURL'] + memoryInitializer;
            }
```
And here is what will be after proposed changes:
```javascript
            var memoryInitializer = '%s';
            if (Module['memoryInitializerPrefixURL']) {
              memoryInitializer = Module['memoryInitializerPrefixURL'] + memoryInitializer;
            } else {
              memoryInitializer = Module['locateFile'](memoryInitializer);
            }
```

This is what I mean by changed priority. Since `Module['locateFile']` is now always defined, we need to lower its priority. We could introduce another function private to Emscripten and call something like `var memoryInitializer = anotherFunction('%s');` seems redundant.

Yes, this is a breaking change for some edge cases. But it doesn't break the code you've instructed other developers with, unless they're also defining other options like `memoryInitializerPrefixURL` at the same time (which is extremely unlikely because it doesn't make any sense).

kripken commented:
For the edge cases, could we show an error if they occur? I don't think we expect `memoryInitializerPrefixURL` to be used with `locateFile`, so perhaps we could just disallow that with a clear error (instead of a potentially surprising behavior change)?

nazar-pc commented:
> I don't think we expect `memoryInitializerPrefixURL` to be used with `locateFile`

Taking into account that with current implementation `locateFile` is always defined (by user or implicitly), this makes `memoryInitializerPrefixURL` obsolete?

If there is no use case where, say, `filePackagePrefixURL` is different from `memoryInitializerPrefixURL`, `cdInitializerPrefixURL` or `filePackagePrefixURL`, then they are all obsolete in favor of generic `scriptDirectory`.

nazar-pc commented:
I think if we encounter any of those `*PrefixURL` we can treat all of them as `scriptDirectory` aliases now.

kripken commented:
Sorry, I wasn't clear. I meant that if the user provides both `memoryInitializerPrefixURL` and `locateFile`, we should give an error. But the user can still provide either one (but just one). (That we internally generate a `locateFile` is an internal detail.)

nazar-pc commented:
Makes sense, will add corresponding checks.

What about deprecating `*PrefixURL`?

juj commented:
> Since Module['locateFile'] is now always defined, we need to lower its priority.

I don't think it's good to always require having `Module['locateFile']` defined. It would be better to move to a direction where we have fewer mandatory properties defined on `Module`, rather than more. The needed path redirection can be implemented without requiring users to create a suitable `Module['locateFile']`.

In particular, the root problem here would still remain, that if user implements their own `Module['locateFile']`, if they return a relative URL from this function, it will still be treated differently in Node.js and in the browser. In Node.js with the proposed change, relative URLs returned by `Module['locateFile']` will still be treated relative to Node.js CWD, and in the browser, relative URLs will be treated relative to the `.html` file. In this form, we are placing an added requirement to developers that they should always return an absolute URL from `Module['locateFile']` and never a relative one, if they want to stay consistent with respect to Node.js and the browser.

> Yes, this is a breaking change for some edge cases. But it doesn't break the code you've instructed other developers with, unless they're also defining other options like memoryInitializerPrefixURL at the same time

This breaks in emunittest suite, where I implement `Module.locateFile` to existing builds that developers provide to override hosting of built projects, in the form of `Module.locateFile: function(url) { return cdnRoot + url + '.gz'; }`. Unity uses both `filePackagePrefixURL` and `memoryInitializerPrefixURL` with either `Release/` or `Compressed/` by default. This change will mean that if one wants to specify `Module.locateFile` to configure paths of all deployed items on a CDN, they'll need to be aware of and delete all the other properties, if they might exist.

> What about deprecating *PrefixURL?

I don't think there is a need to deprecate to fix the issue. These are features that developers are using, and they are tested and there is no inherent problem with them, except than a couple of added bytes. I think we should focus to fix the root problem that relative URLs for these specific files should be treated relative to the script and not the CWD.

I have posted PR #5484 as a suggestion to fix this. It changes the meaning of relative URLs from these built-in functions to be interpreted with respect to the main `.js` file on Node.js, but keep everything else the way it is so it will not break anything.

juj commented on src/shell.js:
```diff
@@ -71,6 +71,27 @@ if (!ENVIRONMENT_IS_PTHREAD) PthreadWorkerInit = {};
+    } else if (ENVIRONMENT_IS_WEB) {
+      Module['scriptDirectory'] = document.currentScript.src.split('/').slice(0, -1).join('/') + '/';
+    } else if (ENVIRONMENT_IS_WORKER) {
+      Module['scriptDirectory'] = self.location.href.split('/').slice(0, -1).join('/') + '/';
```
On the web and in web workers, XHRs are already treated relative to the current location by default, we don't need to explicitly reinforce that.

nazar-pc replied:
http://test/xhr.html:
```html
<!doctype html>
<script src="/xhr/test.js"></script>
```
/xhr/test.js:
```javascript
var xhr = new XMLHttpRequest();
xhr.open('GET', 'test2', true);
xhr.send(null);
```
The call is made to `http://test/test2` instead of `http://test/xhr/test2`.

Also there is `<base>` tag that have influence on relative paths. I think being explicit here is a good thing.

RReverser replied:
I think if someone uses `<base>`, they explicitly want a different base path for any resources and that shouldn't get broken.

nazar-pc replied:
I've just clarified that the statement that was made is not true and there is a value in determining an absolute path here.

nazar-pc commented:
> I don't think it's good to always require having `Module['locateFile']` defined.

It is not required to be specified explicitly by user: https://github.com/kripken/emscripten/pull/5368/files#diff-8f35bc569987bab9a3691d986d90e75fR74

> if they return a relative URL from this function, it will still be treated differently in Node.js and in the browser

This is expected, user have to identify which environment is used and behave accordingly.

> In Node.js with the proposed change, relative URLs returned by `Module['locateFile']` will still be treated relative to Node.js CWD

And why is this a problem? With proposed change most of the time user wouldn't need to touch this option in the first place. And if he/she ever does, I don't see any problem with specifying absolute path.

> and in the browser, relative URLs will be treated relative to the .html file.

In modern applications there might be no single `.html` file and relative paths are more complicated that that (for instance, they also depend on `<base>` tag).

> In this form, we are placing an added requirement to developers that they should always return an absolute URL from Module['locateFile'] and never a relative one, if they want to stay consistent with respect to Node.js and the browser.

This is exactly the same requirement as before. On server especially user should always infer absolute path (using `__dirname` or otherwise) just to be sure that correct path is always used.

> I don't think there is a need to deprecate to fix the issue. These are features that developers are using, and they are tested and there is no inherent problem with them, except than a couple of added bytes.

True, but they are not useful either once there is a universal `Module['scriptDirectory']`. I don't think it is healthy to have 4 prefix variables that literally do the same thing and one more on top of them to act like prefix to already prefixed paths (here I'm talking about #5484). Deprecating those redundant options in advance would give developers plenty of time to migrate, I have nothing against keeping them till next major version.

nazar-pc commented:
I'd like to make this a separate comment to clarify about relative paths:
1. Returning relative paths from `Module['locateFile']` is not a good idea in general (and with proposed change behavior of `Module['locateFile']` stays exactly the same as before, which is good for backward compatibility)
2. Even when relative path is returned, both in Node.js and in browser it will behave exactly like user expects under any other circumstance (e.g. filesystem operations and XHR requests): relatively to CWD, relatively to the root of the website or relatively to the `<base href>`
3. Making relative paths relative to something like script location is unexpected and confusing

nazar-pc commented:
Can we move forward here? It is a pain to apply a patch to latest Emscripten version each time I want to publish newer version of the library to NPM and without this patch build is basically broken.

kripken commented:
Yeah, we should really resolve this. Reading this and the parallel #5484, it looks like there is more consensus on the other one - @nazar-pc, would merging that other one be good enough for you?

nazar-pc commented:
No. Primarily because #5484 it doesn't solve the issue for browser at all. Secondarily because even if it fixes the issue for browser it will take more bytes in JavaScript build in long term.

kripken commented:
I see. Well, I think only you and @juj really understand this, so I hope you two can figure out the right solution. If not I can find some time to study the issue more myself.

curiousdannii commented:
My impression is that a lot of the difficulty is maintaining compatibility, and how best to do so. If there was going to be a time soon for a major breaking release then these problems could be solved simpler.

juj commented:
> No. Primarily because #5484 it doesn't solve the issue for browser at all.

I still think the browser specific issue is not really much of an issue.

The difference between node.js case and browser case is that in node.js, one can `cd` to another directory, and then attempt to `node ../../path/to/compiled/code/main.js` which is something that is expected to work on any command line programs. Fixing that side is all agreed.

In the browser, content is loaded relative to the output html file. The files `output.html` and `output.js` are located in the same directory on CDN, it works. The example that was posted in https://github.com/kripken/emscripten/pull/5368#discussion_r133534321 illustrate a scenario where a developer deliberately uploads files `output.html` and `output.js` to different paths on a CDN. In that kind of scenario where one explicitly copies output files to different directories on a web server, I don't think it is unreasonable to configure to point to the different download location paths.

Also, I don't have data to say whether loading the relative `.js` files with respect to `<base href>` is a desired feature or a bug. It could be what developer wants, or then not. One will, after all, have to deliberately split the output files to different directories for a discrepancy to happen here.

The problem why this PR is not acceptable is that it changes the semantics of the existing methods, which would be a breaking change for not more worth than peace of mind/simplicity. We can not change semantics on those grounds. Also, I think this PR does not pass test https://github.com/kripken/emscripten/pull/5526, because `document.currentScript.src` may refer to a blob URL that one cannot infer relative paths from.

> Secondarily because even if it fixes the issue for browser it will take more bytes in JavaScript build in long term.

Regarding the size increase from added code size, in this comment https://github.com/kripken/emscripten/pull/5484#issuecomment-325195387 you seemed to agree that would be ok, and the proposed path to slim down would be to start ifdeffing out the unneeded features would be alright. Though it has been some time since that comment, so it is ok if you have a different opinion since then. In terms of immediate added code size, I think #5484 is very comparable, if not a bit smaller, than this PR.

nazar-pc commented:
> In the browser, content is loaded relative to the output html file

I'm not sure you're developing for the web currently, but in case you don't I'd like to clarify that in modern days almost no one will use HTML file generated by Emscipten for anything more than a demo. Nowadays we use lots of libraries that are located all over the place.

So there is no such thing like `output html file`, instead there are a lot of dynamically generated pages and different pages load different set of dependencies (some of which may be Emscripten modules) from different places. So Emscripten should not make it painful to work in such environment, but it currently does.

> Also, I don't have data to say whether loading the relative .js files with respect to `<base href>` is a desired feature or a bug.

`.js` file loads and works great already. The question here is whether `.wasm` or `.mem` file that is located in the same directory as `.js` file will be loaded successfully. With this PR yes, it will be loaded without issues regardless of `<base href>` and `.js` file location. In #5484 you assume that problem only exists in Node.js, which is not true, and thus `.wasm` and `.mem` files will often fail to load in browser.

> The problem why this PR is not acceptable is that it changes the semantics of the existing methods, which would be a breaking change for not more worth than peace of mind/simplicity. We can not change semantics on those grounds.

I agree about slight semantics change, but disagree that it is a significant breaking change. Could you provide a single real-world example where this will become an issue? Because I truly believe no one will notice anything different after upgrade, would be happy to adjust PR if I'm wrong on this.

> Also, I think this PR does not pass test #5526

You are right, added one more commit with a fix and rebased against incoming: test passes now.

> Regarding the size increase from added code size, in this comment #5484 (comment) you seemed to agree that would be ok

Yes, this is not my primary concern at the moment, let's skip its discussion for now.

juj commented:
> I'm not sure you're developing for the web currently, but ...

Alright, this is now veering off the target. Commented in https://github.com/kripken/emscripten/pull/5484#issuecomment-336257669 for the next steps.

curiousdannii commented:
> It's important to note that this will not reference the <script> element if the code in the script is being called as a callback or event handler; it will only reference the element while it's initially being processed. ([MDN on document.currentScript](https://developer.mozilla.org/en-US/docs/Web/API/Document/currentScript))

Does that mean it this pull request won't work if you use the MODULARIZE `.then()` feature?

nazar-pc commented:
> it will only reference the element while it's initially being processed

This is exactly when it is used, this PR doesn't break anything in that regard. I'm already using `MODULARIZE=1` and `.then()` in my projects with this PR applied.

curiousdannii commented:
@nazar-pc Good to hear. I thought `.then` might delay the processing.

nazar-pc commented:
Merged incoming, fixed single failed test.
CI is green (build in my fork finished already: https://travis-ci.org/nazar-pc/emscripten/builds/305144046).
I hope this can finally be merged.

nazar-pc commented:
@kripken, @juj, what can I do to get this merged sooner than later?

kripken commented:
We need @juj's feedback here. On vacation currently, when he's back we should definitely try to resolve this quickly, I agree it's been going on too slowly for too long.

Hopefully we can resolve this next week when he's back.

juj commented on emcc.py:
```diff
@@ -2493,7 +2493,7 @@ def generate_html(target, options, js_target, target_basename,
           if (typeof Module['locateFile'] === 'function') {
             memoryInitializer = Module['locateFile'](memoryInitializer);
-          } else if (Module['memoryInitializerPrefixURL']) {
+          } else {
```
We don't want to introduce the requirement that these variables would exist unconditionally in `Module`. If they don't, then this would be adding up `undefined` + string.

nazar-pc replied:
Code in shell.js ensures that either `Module.locateFile` or `Module.memoryInitializerPrefixURL` is present without semantic change, so this condition is simply redundant.

juj replied:
The role of the `Module` object has usually been a way for users to add fields to do customization of the loading experience. We have generally had few mandatory fields present there that would cause a build to fail to load. I still think it's better to solve this routing via a mechanism that has less state to it, i.e. just add the single `Module['scriptDirectory']` variable.

Doing this kind of change where we begin to unconditionally depend on `Module['memoryInitializerPrefixURL']` takes the loader code to a direction where it is more dependent on this variable, rather than less, so I think the approach in #5484 is better.

nazar-pc replied:
@juj, have you seen that `Module['memoryInitializerPrefixURL']` fallbacks to just `Module['scriptDirectory']`?
`Module['scriptDirectory']` on the other hand have solid defaults and you don't need to define it unless you need to change something.

juj replied:
Right - I still think it's better to rely less on fields being mandatorily present in `Module`, and rather have an initialization flow that does not require e.g. `Module['memoryInitializerPrefixURL']` always to exist. I know it is initialized unconditionally in the loader, but that does cost code lines, and the initialization flow is asynchronous, so things may get changed in between. It seems more refactor-hardened to me to not require that the older `Module['*PrefixURL']` need to unconditionally exist.

nazar-pc replied:
Those properties are assigned synchronously on Module initialization before any use, so fortunately asynchronicity is not an issue in this case.

> but that does cost code lines

This can be refactored later, most of the conditional statements will go away when `Module['*PrefixURL']` is dropped somewhere in future. It will boil down to single Closure-compressable property for everything.

juj commented on src/Fetch.js:
```diff
@@ -114,7 +114,7 @@ var Fetch = {
       // of these are passed, then the default URL 'pthread-main.js' relative to the main html file is loaded.
       if (typeof Module['locateFile'] === 'function') fetchJs = Module['locateFile'](fetchJs);
-      else if (Module['pthreadMainPrefixURL']) fetchJs = Module['pthreadMainPrefixURL'] + fetchJs;
+      else fetchJs = Module['pthreadMainPrefixURL'] + fetchJs;
```
Here also we can get `undefined`+'someString'`, and we don't want to require the URL variable to be present.

nazar-pc replied:
Same here, either `Module.locateFile` or `Module.pthreadMainPrefixURL` will be present, condition is redundant

juj commented on src/preamble.js:
```diff
@@ -2034,6 +2034,10 @@ function integrateWasmJS() {
+  } else {
+    wasmTextFile = Module['scriptDirectory'] + wasmTextFile;
+    wasmBinaryFile = Module['scriptDirectory'] + wasmBinaryFile;
+    asmjsCodeFile = Module['scriptDirectory'] + asmjsCodeFile;
```
Here `scriptDirectory` is applied as an alternative to `locateFile`. I think a good semantic would be that `scriptDirectory` would be applied on top of `locateFile` or the other prefixes. That way if one prepends a relative URL from a `locateFile` function, e.g. `Module.locateFile = function(url) { return 'Release/' + url' }`, then `scriptDirectory` will still be applied to it, producing the CWD routing result.

nazar-pc replied:
This is exactly the same semantic that was used in other places before with options like `memoryInitializerPrefixURL`.

Chaining these options makes things more complex and increases cognitive load when you're trying to figure out what the final path would be. For such a niche use case I don't think over engineering makes a lot of sense.

juj replied:
> Chaining these options makes things more complex and increases cognitive load

Returning a relative path from `Module.locateFile()` is a use case that has worked before, and with code like above, if one defined `Module.locateFile()`, it would then not receive the CWD fix for Node.js. Treating relative return values from `Module.locateFile()` with respect to `Module['scriptDirectory']` will allow those use cases to also work.

nazar-pc replied:
> if one defined Module.locateFile(), it would then not receive the CWD fix for Node.js

Because it is not supposed to! I don't really see why you'd want some additional tweaks to be added to the result of location function that you've specified. It doesn't make any sense to me.

juj commented on src/shell.js:
```diff
@@ -72,6 +72,37 @@ if (!ENVIRONMENT_IS_PTHREAD) PthreadWorkerInit = {};
+    Module['cdInitializerPrefixURL'] = Module['scriptDirectory'];
+  }
+  if (!Module['filePackagePrefixURL']) {
+    Module['filePackagePrefixURL'] = Module['scriptDirectory'];
```
Mixing these seems unnecessary, especially if we're working towards removing them in the future - less code lines the better.

nazar-pc replied:
Similar conditions were present before already, they just were places in various locations, so code size does not increase much. In future we can remove these lines and replace every occurrence of `*PrefixURL` with `scriptDirectory`. I'm open for suggestions on cleaner ways to do this.

juj commented:
> @kripken, @juj, what can I do to get this merged sooner than later?

Apologies for the delays. The situation is still much the same as commented in https://github.com/kripken/emscripten/pull/5484#issuecomment-336257669.

juj commented:
I still think PR https://github.com/kripken/emscripten/pull/5484 is the better way to solve this. It makes all relative paths in Node.js to be treated relative to `Module['scriptDirectory']` (`== __dirname` if not otherwise specified), instead of user's CWD. Loading relative to CWD is behavior that nobody ever wanted, so that is not contentious, and can be fixed immediately.

The PR also retains all relative paths in browser to be relative to baseURI, i.e `<base>`, or the location of the `.html` file. Changing this behavior is more contentious, because users have already utilized relative URLs in *PrefixURLs and Module.locateFile(), so if those relative URLs were changed to load relative to the main .js file instead of the main .html file, it will break users, and it should be treated separately, and with a way that does allow developers to keep loading relative to `<base>` if desired. That is why such a change should be approached separately after #5484.

curiousdannii commented:
We've been talking about a major release, so I think it would help to think what the ideal would be if backwards compatibility was not a concern.

If I'm reading the room correctly there are two main desires:

1. Consolidation of path settings. IMO the ideal would be one setting, `scriptDirectory`, and one function, `locateFile`. `locateFile` would really only be for people who want to for some reason store files in different places.

    `scriptDirectory` can be added without breaking backwards compatibility. When it's time for a major release, another PR could remove and clean up the other settings.

2. Sensible defaults. Ideally 95%+ of users wouldn't have to set either `scriptDirectory` or `locateFile`.

    #4942 has been languishing for almost a year now. There's no good option for browsers currently either. We will want reliable defaults based on `__dirname`. `document.currentScript`, `self.location` etc, and then some docs to explain what to do if you're in the then unusual situation of needing to change it manually.

    (Side node for Node: I'm not sure if it's possible, but we might be able to write out code so that browserify and other tools won't detect and transform `__dirname`. In the future we'll also have to support ES Modules too.)

juj commented:
> This is a flaw in your approach. My PR is not affected simply because it doesn't change the root of the relative paths at all, even on Node.js, it only affects default values.

If only the defaults are changed, it means that developers will need to bake in the same `__dirname` aware fix to their code if they need to change `Module.locateFile()`. It is better to define the semantics of relative paths for `Module.locateFile()` directly so that even when you do need to customize the function, the relative paths get loaded w.r.t. the JS file and not w.r.t to CWD.

juj commented:
> Consolidation of path settings. IMO the ideal would be one setting, scriptDirectory, and one function, locateFile

This looks like is agreed on. The discrepancy seems to come from what the semantics should be when these two are defined at the same time. E.g. what happens with the following?

```js
Module['scriptDirectory'] = ENVIRONMENT_IS_NODE ? __dirname : '';
Module['locateFile'] = function(url) { return bIsCompressed ? ('Compressed/'+url+'.gz') : ('Development/'+url); }
```

(the above style of `locateFile` function is being used in Unreal Engine 4 for example, where the function is used to route to gzip-compressed variants of assets if compression is enabled)

My proposal is that relative paths returned from `Module['locateFile']` would be treated relative to `Module['scriptDirectory']`, @nazar-pc is suggesting that if a developer defines `Module['locateFile']`, then `Module['scriptDirectory']` will be ignored.

> Sensible defaults. Ideally 95%+ of users wouldn't have to set either scriptDirectory or locateFile.

I understand this is the same in all proposals. I.e. default to 

```js
Module['scriptDirectory'] = ENVIRONMENT_IS_NODE ? __dirname : '';
Module['locateFile'] = undefined;
```

to fix up the Node.js CWD vs rooted-to-.js-root behavior.

nazar-pc commented:
@juj, what if we pass `Module.scriptDirectory` (either resolved by Emscripten or developer-specified) as second argument to `Module.locateFile()`? It will not cause any breaking change to existing code and developer can use it if needed.

I think this will check all boxes with no breaking changes.

nazar-pc commented:
Though, developer can kind of do this already (I'll need to update this PR slightly)

nazar-pc commented:
> Where does the confusion come from, and how does it break?

There are relative paths in CSS and there are JavaScript Modules - these are language constructs and are resolved relatively to the file they are used in. All other relative paths, including Fetch and XHR follow different rules and Emscripten should better follow them rather than using something else.

If someone uses relative paths in `Module.locateFile()` already, additional tweaks on top of that will break that code.

With latest commit you can do something as simple as:
```javascript
Module.locateFile = function(url) {
  return Module.scriptDirectory + '/dir/' + url;
}
```

Is this enough?

curiousdannii commented:
@juj I was going to ask what the use case for supporting a multi platform locateFile would be, but you provided a good example.

Passing scriptDirectory to locateFile makes lots of sense. In simple cases, or where you're in charge of the absolute paths, it can be ignored. Unreal Engine would have to update their code, but it wouldn't be too complex.

curiousdannii commented:
@nazar-pc I do prefer your idea of passing in scriptDirectory, because it allows you to define your locateFile function even if your module isn't in scope for some reason.

nazar-pc commented:
@curiousdannii, you can use `this['scriptDirectory']` as it is assigned as property on `Module` anyway.

curiousdannii commented:
Oh yeah, of course. No need for an extra parameter then.

nazar-pc commented:
I've put some final tweaks yesterday to address the issues that appeared when `-s MODULARIZE=1` is used in browser environment and confirmed on 4 of my projects that it does work well now.
Tests are also green and I believe all of the questions were answered and all of the mentioned concerns were addressed.

@kripken, @juj, this PR is open for over half a year, can we make a cooperative effort to get this merged before next release?

kripken commented:
I agree we should make an effort to get this resolved, yes.

For my part, I haven't been following all the details of the complex conversation here, and at this point it is great deal to read and understand all of it. Would it be possible to briefly summarize where the PR now stands: what it does, what is fixed, and tradeoffs if any?

nazar-pc commented:
# Contents of this PR as of right now

* added generic `Module.scriptDirectory` option that specifies an absolute path to directory from which `.wasm`, `.mem` and other files are loaded
* `Module.scriptDirectory` defaults to the directory where module's JS file is located, this is the way Emscripten produces build results and the most reasonable one
* `Module.scriptDirectory` is used as fallback for old `*PrefixURL` options; it doesn't replace old options if they are explicitly specified, hence backwards compatible; allows for easy removal of `*PrefixURL` options when and if there will be a need to do this
* thanks to `Module.scriptDirectory` and its defaults, `.wasm`, `.mem` and other files are loaded correctly even when moved somewhere to subdirectories or even CDN both in Browser and Node.js
* there are no backward incompatible changes that may affect existing users as far as I'm aware of, everything ever mentioned in this thread was addressed

I'm already using Emscripten version from this PR for [5 of my NPM packages](https://github.com/nazar-pc?tab=repositories&q=wasm), they were tested in various conditions both in browser and on server and all work great so far.

paulshapiro commented:
Would love to see this get merged.

paulshapiro commented:
Looks like a trivial rebase is necessary here, but the methodology seems sound to me, and merging it would give us a usable node.js (used in e.g. continuous testing) build gives us the chance to iterate off this fix and refine specific parameter names/arrangements as we go.

nazar-pc commented:
I've rebased this a few times already, will do again once there is an intention for merge

paulshapiro commented:
@juj @kripken I think we could merge this, couldn't we?

paulshapiro commented:
@nazar-pc one thing I think would be really helpful to merging here is if you were to squash your relevant commits so that it's a smooth and clear patch.

kripken commented:
@nazar-pc First thing, I apologize for not getting to this sooner. Ok, I spent some time to read this in detail now. I think this is in the right direction, and we should really get this in. I'd actually suggest something that goes a step further, though (I see related things were discussed in the past here), how about this:

 * We *remove* the `Module.*Prefix` API. As with other deprecated APIs, we can show a detailed error message in builds with `ASSERTIONS` (which includes default `-O0`). This lets us remove a bunch of unnecessary JS, and also avoid adding more JS for backwards compatibility in this PR.
 * The only user-facing API we support is `Module.locateFile`.
 * Instead of `Module.scriptDirectory = "dir/";`, a user would do `Module.locateFile = function(x) { return "dir/" + x };` - does that make sense, or have I missed something?
 * The new functionality in this PR, of autodetecting the script dir location using `.location` etc., is done if `Module.locateFile` is not defined by the user.

What do you think?

If that sounds good in general, I could do the first bullet point myself if you prefer. Again, sorry this PR has taken so long to resolve.

Btw, one question about the core here, I see these comments:
```
We do this for `-s MODULARIZE=1` where at the time when this code runs self.location might already be unavailable
```
What would make `self.location` etc. unavailable?

kripken commented:
Yeah, @juj was more concerned about backwards compatibility. Which is reasonable, but I think in recent months we've had a good experience with removing APIs together with good errors in ASSERTION builds - it's still slightly annoying for users, but it's a quick fix for them. So @juj might feel differently now. But regardless this also comes down to who does maintenance in the project here, and @juj has less time for that now. Personally I'm confident we can maintain things going forward with these changes, so we can go ahead.

About `window.location`, thinking some more I guess `window.location` can change as more scripts are loaded on the page, or the page redirects itself, etc.? And in MODULARIZE (but not MODULARIZE_INSTANCE) there is more time for that to happen, as the user may create the instance later. But I'm not sure when/how the location property changes. Yeah, will be good to see if a test fails.

nazar-pc commented:
Wasn't following Emscripten development recently, great to see things are moving quicker.
I have less time for Open Source development too, but I'm pretty sure you can expect me to address above suggestions and merge `incoming` back into this PR again on Saturday or Sunday.

curiousdannii commented:
I'd prefer to keep it as `scriptDirectory`. (Or another string property with a different name if anyone thinks it's confusing.) `locateFile` should be the power user option for people who need to store files in different places. Regular users should just be able to pass in a path. They shouldn't be required to give a function that assembles the full paths. They shouldn't need to remember if there's a trailing slash.

kripken commented:
@curiousdannii I'm not strongly opposed if there is interest in a simple version of `locateFile`. (Maybe `filePrefix` instead of `scriptDirectory`?) Seems redundant to me but I'm ok with it if more people feel otherwise.

Not sure what you mean by the trailing slash, though. It seems like regardless users need to know whether to include or not include one? (Also, I think an extra slash is not an error in URLs?)

nazar-pc commented:
There was an alternative suggestion in earlier discussions that I've implemented back then in this branch: https://github.com/nazar-pc/emscripten/compare/fix-wasm-loading-for-moved-files...nazar-pc:fix-wasm-loading-for-moved-files-locateFile-second-argument
This way we can remove `scriptDirectory` as separate option, but still pass its auto-detected value into user-defined `locateFile`.

It should satisfy both @kripken and @curiousdannii, what do you guys think?

nazar-pc commented:
All tests are good except `test_glgears_proxy_jstarget`, which works fine in Firefox, but fails in Chromium Stable/Nightly even on `incoming` branch (tested both locally) despite with different error.

Other than that tests are green and everything should be ready.

Just to make it easier to follow, updated state of PR:

# Contents of this PR as of right now

* `Module.locateFile()` resolves files relatively to JavaScript file's directory rather than current working directory in Node.js and similar path in browser (shouldn't break existing users)
* `Module.locateFile()` now has second argument `prefix` that is JavaScript file's directory with trailing `/`, can be useful when it is necessary to load different files from different places for some reason (shouldn't break existing users either)
* `Module.*PrefixURL` options removed, added errors in ASSERTION builds for those

paulshapiro commented:
Reviewed the PR and would be happy to help test this (although currently stuck on how to point emsdk to another repo's branches).

nazar-pc commented:
@paulshapiro, I've just symlinked Emscripten from cloned repo as `emcc_dev` and calling it as is. Works fine for my purposes. You don't need to change anything in emsdk unless testing older versions of Emscripten with older clang or other components.

nazar-pc commented on site/source/docs/api_reference/module.rst:
```diff
@@ -51,15 +51,9 @@ The following ``Module`` attributes affect code execution. Set them to customize
 .. js:attribute:: Module.locateFile
 
-	If set, this method will be called when the runtime needs to load a file, such as a ``.wasm`` WebAssembly file, ``.mem`` memory init file, or a file generated by the file packager. The function receives the URL, and should return the actual URL. This lets you host file packages or the ``.mem`` file etc. on a different location than the current directory (which is the default expectation), for ex ...
+	If set, this method will be called when the runtime needs to load a file, such as a ``.wasm`` WebAssembly file, ``.mem`` memory init file, or a file generated by the file packager. The function receives the relative path to the file as configured in build process and a prefix (path to JavaScript file), and should return the actual URL. This lets you host file packages or the ``.mem`` file etc. o ...
```
Yes, there are some places where `Module['locateFile']` is called, but `Module` itself was not initialized yet. I'd like to see this re-designed this somehow, but not sure if it is possible.

I've added a note in latest commits about these cases to the degree I understand them.

nazar-pc commented on src/shell.js:
```diff
@@ -83,6 +83,40 @@ if (!ENVIRONMENT_IS_PTHREAD) PthreadWorkerInit = {};
+#endif
+// `/` should be present at the end if `scriptDirectory` is not empty
+var scriptDirectory = '';
+#if !SUPPORT_BASE64_EMBEDDING
```
I didn't fully understand how it works when `SUPPORT_BASE64_EMBEDDING` landed, but it seems like there were some guards added in relevant places, I'll try to remove this check and see if any tests fail.

nazar-pc commented on src/shell.js:
```diff
@@ -83,6 +83,40 @@ if (!ENVIRONMENT_IS_PTHREAD) PthreadWorkerInit = {};
+#if !SUPPORT_BASE64_EMBEDDING
+#if ENVIRONMENT_MAY_BE_NODE
+if (ENVIRONMENT_IS_NODE) {
+  scriptDirectory = __dirname + '/';
```
It can be, but I'd prefer to keep everything related to `Module._locateFile` next to each other, this way it is easier to follow execution flow.

kripken replied:
It's easier in that sense, I agree, but it's less consistent with the style above, which is one section for web, one for node, etc. So I do see your point, but I think it's more important to keep the overall consistency in the file. (In addition, it's also shorter that way.)

nazar-pc replied:
OK, will refactor it.

nazar-pc commented on src/shell.js:
```diff
@@ -83,6 +83,40 @@ if (!ENVIRONMENT_IS_PTHREAD) PthreadWorkerInit = {};
+}
+#endif
+#endif // SUPPORT_BASE64_EMBEDDING
+Module._locateFile = function (path) {
```
I think we can, applied this change in one of latest commits.

nazar-pc commented:
Now different test `test_sdl2_image_prepare` failed instead, looks like some flakiness on Circle CI's side. Works fine for me locally.

paulshapiro commented:
@nazar-pc testing under Electron in the renderer (web) process, I don't believe it thinks it's node.js, so it looks for the file in the dir that holds the index.html file. Any advice?

nazar-pc commented:
@paulshapiro, if you're building module just for Electron, then try `ENVIRONMENT` from https://kripken.github.io/emscripten-site/docs/optimizing/Optimizing-Code.html#miscellaneous-code-size-tips or otherwise this is one of the use cases where you actually need to define custom `Module.locateFile`.

paulshapiro commented:
@nazar-pc that's just the thing - our lib is consumed by people using all kinds of different platforms - web, Node.JS, React, etc. We've had a working emscripten build for ~4 yrs but it's getting modernized now. I was thinking this is the use-case for locateFile but my concern is that given there's a `require()` (`MODULARIZE` on), I can't specify parameters to the module load beforehand. Feel like I'm missing something :)

nazar-pc commented:
There was `Module['ENVIRONMENT']` before that allowed you to force specific environment overriding autodetection. I'd like to avoid introducing workarounds for Electron specifically.
In the mean time it might be possible to do by specifying custom library.

paulshapiro commented:
@nazar-pc @curiousdannii fwiw here's what I'd been using

`var ENVIRONMENT_IS_WEB=typeof window==="object";
var ENVIRONMENT_IS_WORKER=typeof importScripts==="function";
var ENVIRONMENT_IS_NODE=typeof process==="object"&&process.browser !== true&&typeof require==="function" && ENVIRONMENT_IS_WORKER == false; // we want this to be true for Electron but not for Cordova nor a WebView
var ENVIRONMENT_IS_SHELL=!ENVIRONMENT_IS_WEB&&!ENVIRONMENT_IS_NODE&&!ENVIRONMENT_IS_WORKER;`

kripken commented on emcc.py:
```diff
@@ -2598,7 +2598,10 @@ def modularize():
+};
+%(EXPORT_NAME)s = %(EXPORT_NAME)s.bind({
+  _currentScript: typeof document !== 'undefined' ? document.currentScript : undefined
+})%(instantiate)s;
```
What do you think about leaving this up to the user, instead? We could document that `document.currentScript` is used, and so if the page will modify that, the user should implement `locateFile` in order to get things working properly.

Alternatively, we could make more of an effort here: write up tests that ensure the right behavior, and document what cases exactly we handle. Basically, it's a "best-effort" without clear guidelines and tests that I'd prefer to avoid.

nazar-pc replied:
Well, my idea was to make it just work by default and it does. I'd like to avoid user to require doing anything special just to get this basic thing working, re-implementing `locateFile` would be a big pain, especially since it would be necessary to get script's `document.currentScript` from the outside.

I'm more inclined to write more tests here rather than dealing with it every time I use Emscripten.

kripken commented on emcc.py:
```diff
@@ -2704,11 +2707,7 @@ def generate_html(target, options, js_target, target_basename,
-          } else if (Module['memoryInitializerPrefixURL']) {
-            memoryInitializer = Module['memoryInitializerPrefixURL'] + memoryInitializer;
-          }
+          memoryInitializer = Module['locateFile'] ? Module['locateFile'](memoryInitializer, '') : memoryInitializer;
```
I see, right, this is code in the HTML, which declares `Module`, but the details are filled in later in the JS. So this is good.

kripken commented:
About Electron, #6803 has a proposal for users replacing the "core JS" which is what detects and sets up the environment. That could provide a way to define a custom Electron core JS file.

In the shorter term, if Electron is currently blocked, we could add an option to do something like the old `Module['ENVIRONMENT']` runtime detection, but let's open a separate issue for that discussion please.

paulshapiro commented:
> In the shorter term, if Electron is currently blocked

@kripken understood, thanks. The reason I'm asking about it in this venue is that if the parent directory behavior for file location is contingent upon some set of environment detection functions which are internal to Module, then they're only being used indirectly to trigger certain expected loading functionality in emscripten. That loading behavior may not actually be well defined anymore given that Node can be run in so many contexts. Reproducing the same loading behavior would ostensibly be done by mapping the actual environment to the one which will have the desired loading behavior, but that causes complexity because the environment might not truly be only node. For that reason it leaves me to wonder if feature detection is not a more fitting method in the sense that a desired parent directory or locate file function could be passed in as desired, relieving emscripten of the burden to match environments.

nazar-pc commented:
@paulshapiro, I think these kind of issues should be discussed separately. You can get current behavior of Emscripten after this PR by simply using `Module.locateFile = function (path) {return path;}`.
So at least it doesn't get worse than before. While I'd like to see Electron and other environments supported, let's merge at least something eventually.

If there are other issues to be fixed here, let me know and I'll be able to resolve them on weekend, otherwise I think PR is in pretty good shape.

nazar-pc commented:
@kripken, addressed your last comments in 2 latest commits.

kripken commented:
Great, thanks @nazar-pc!

Let's merge this in now. I have a few minor changes, mainly in docs, that I'll do in a followup PR to this one.

I just had one question about the code (that can also wait for a followup PR, if we need one) - I see you handle a `blob:` URL for the web but not worker case - is there a reason not to also check in the worker case as well? It would also be nice to have a test for the blob URL case.

Thanks again for your patience in this PR - sorry it took this long to get merged.

nazar-pc commented:
I believe there is already a test case for `blob:` that was failing with earlier iterations of this PR, so I've added corresponding check.
It is hard to imagine why something like that would be useful in worker though, but in case someone have a legitimate use case it is always possible to add one more check there.

RReverser commented:
Omg, thank you @nazar-pc and @kripken. I lost hope that this would be merged a long time ago, happy to be proved wrong.

brion commented:
Hmm, that 'bind' call for setting _currentSource doesn't appear to do anything. With 1.38.10 with this patch, ogv.js's demuxer modules are broken because this._currentSource is null when checked. https://github.com/kripken/emscripten/issues/6903

brion commented:
Ok, that was breaking because I was calling the module function with "new", which replaces the "this" objected provided by the "bind" with a new one with no properties on it.

Fix is to follow current documentation and not use "new" with the module function (it's not a constructor, but an .... initializer?)
